File: android-app/src/main/java/fr/cph/chicago/data/Preferences.java
package fr.cph.chicago.data;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.Date;
import java.util.List;

import fr.cph.chicago.entity.enumeration.TrainDirection;
import fr.cph.chicago.entity.enumeration.TrainLine;

public interface Preferences {

    void saveTrainFavorites(Context context, List<Integer> favorites);

    boolean hasFavorites(Context context);

    void saveBikeFavorites(Context context, List<String> favorites);

    @NonNull
    List<String> getBikeFavorites(Context context);

    void addBikeRouteNameMapping(Context context, String bikeId, String bikeName);

    @Nullable
    String getBikeRouteNameMapping(Context context, String bikeId);

    void saveBusFavorites(Context context, List<String> favorites);

    @NonNull
    List<String> getBusFavorites(Context context);

    void addBusRouteNameMapping(Context context, String busStopId, String routeName);

    @Nullable
    String getBusRouteNameMapping(Context context, String busStopId);

    void addBusStopNameMapping(Context context, String busStopId, String stopName);

    @Nullable
    String getBusStopNameMapping(Context context, String busStopId);

    @NonNull
    List<Integer> getTrainFavorites(Context context);

    void saveTrainFilter(Context context, Integer stationId, TrainLine line, TrainDirection direction, boolean value);

    boolean getTrainFilter(Context context, Integer stationId, TrainLine line, TrainDirection direction);

    Date getRateLastSeen(Context context);

    void setRateLastSeen(Context context);

    void clearPreferences(final Context context);
}

File: android-app/src/main/kotlin/fr/cph/chicago/data/Preferences.kt
package fr.cph.chicago.data

import android.content.Context

import java.util.Date

import fr.cph.chicago.entity.enumeration.TrainDirection
import fr.cph.chicago.entity.enumeration.TrainLine

interface Preferences {

    fun saveTrainFavorites(context: Context, favorites: List<Int>)

    fun hasFavorites(context: Context): Boolean

    fun saveBikeFavorites(context: Context, favorites: List<String>)

    fun getBikeFavorites(context: Context): MutableList<String>

    fun addBikeRouteNameMapping(context: Context, bikeId: String, bikeName: String)

    fun getBikeRouteNameMapping(context: Context, bikeId: String): String?

    fun saveBusFavorites(context: Context, favorites: List<String>)

    fun getBusFavorites(context: Context): MutableList<String>

    fun addBusRouteNameMapping(context: Context, busStopId: String, routeName: String)

    fun getBusRouteNameMapping(context: Context, busStopId: String): String?

    fun addBusStopNameMapping(context: Context, busStopId: String, stopName: String)

    fun getBusStopNameMapping(context: Context, busStopId: String): String?

    fun getTrainFavorites(context: Context): MutableList<Int>

    fun saveTrainFilter(context: Context, stationId: Int, line: TrainLine, direction: TrainDirection, value: Boolean)

    fun getTrainFilter(context: Context, stationId: Int, line: TrainLine, direction: TrainDirection): Boolean

    fun getRateLastSeen(context: Context): Date

    fun setRateLastSeen(context: Context)

    fun clearPreferences(context: Context)
}
